### PR TITLE
Feature/jepa

### PR DESCRIPTION
**Pull request by benjijamorris**
## What does this PR do?
Add Joint Embedding Predictive Architecture and Image world Model (draft) infrastructure.

Fixes #\<issue_number>

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **test your PR locally** with `pytest` command?
- [x] Did you **run pre-commit hooks** with `pre-commit run -a` command?

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: cyto_dl/models/jepa/jepa_base.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        target_embeddings = rearrange(target_embeddings, "b t c -> t b c")
+        target_embeddings = take_indexes(target_embeddings, mask)
+        target_embeddings = rearrange(target_embeddings, "t b c -> b t c")
+        return target_embeddings
+
+    def get_context_embeddings(self, x, mask):
+        # mask context pre-embedding to prevent leakage of target information
+        context_patches, _, _, _ = self.encoder.patchify(x, 0)
+        context_patches = take_indexes(context_patches, mask)
+        context_patches = self.encoder.transformer_forward(context_patches)
```
Comment (ritvikvasan) [inline cyto_dl/models/jepa/jepa_base.py:103]: can you replace transformer_foward here with regular forward for simplicity? the only difference here is applying the mask to the context patches right?

File: cyto_dl/nn/vits/jepa.py [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+            mask = transformer(mask, context_emb)
+        # norm and project back to input dimension
+        mask = self.projector_embed(self.norm(mask))
+        return mask
+
+
+from monai.networks.blocks import UnetOutBlock, UnetResBlock
+
+
+class JEPASeg(torch.nn.Module):
```
Comment (ritvikvasan) [inline cyto_dl/nn/vits/jepa.py:137]: why is this class called Seg? Is this for segmentation only?

File: cyto_dl/nn/vits/jepa.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+            ),
+        )
+
+    def forward(self, img):
+        features = self.encoder(img)
+        features = self.decoder(features)
+        return features
+
+
+class IWMPredictor(torch.nn.Module):
```
Comment (ritvikvasan) [inline cyto_dl/nn/vits/jepa.py:248]: is this class used anywhere?

**Review by ritvikvasan — APPROVED** [review]
This is a lot of work!! Awesome stuff. I had some clarification comments mainly. Very curious how the label-free JEPA did

File: cyto_dl/models/jepa/jepa_base.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            + (1 - self.hparams.momentum) * self.current_epoch / self.hparams.max_epochs
+        )
+
+    def update_teacher(self):
+        # ema of teacher
+        momentum = self._get_momentum()
+        # momentum update of the parameters of the teacher network
+        with torch.no_grad():
+            for param_q, param_k in zip(self.encoder.parameters(), self.teacher.parameters()):
+                param_k.data.mul_(momentum).add_((1 - momentum) * param_q.detach().data)
```
[the remark this answers is not in the document]
Comment (benjijamorris) [reply cyto_dl/models/jepa/jepa_base.py:88]: good catch, updated!

File: cyto_dl/nn/vits/jepa.py (hunk 3 above)
Comment (benjijamorris) [reply cyto_dl/nn/vits/jepa.py:248]: yes, it's passed to the IWM class. I'll add configs to show how all the pieces come together

File: cyto_dl/models/jepa/iwm.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+import pandas as pd
+import torch
+import torch.nn as nn
+from einops import rearrange, repeat
+
+from cyto_dl.models.jepa import JEPABase
+
+
+class IWM(JEPABase):
```
[the remark this answers is not in the document]
Comment (benjijamorris) [reply cyto_dl/models/jepa/iwm.py:11]: True. I'll generalize this to any paired image translation and make that more clear in the docs.

File: cyto_dl/models/jepa/jepa_base.py (hunk 1 above)
Comment (benjijamorris) [reply cyto_dl/models/jepa/jepa_base.py:103]: as it's written now, using regular forward would leak target information into the context. Instead of having transformer_forward and forward as separate functions, I could just add a `patchify` flag to the forward pass and set it to false here?

Comment (ritvikvasan) [reply cyto_dl/models/jepa/jepa_base.py:103]: this makes sense

File: cyto_dl/image/transforms/generate_jepa_masks.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            # randomly select block shape
+            aspect_ratio = self.R.uniform(*self.block_aspect_ratio)
+            width = int(self.mask_size * aspect_ratio)
+            height = int(self.mask_size / aspect_ratio)
+            # randomly select block position
+            x = self.R.randint(0, self.num_patches[-1] - width + 1)
+            y = self.R.randint(0, self.num_patches[-2] - height + 1)
+            # add block to mask
+            if self.spatial_dims == 3:
+                mask[:, y : y + height, x : x + width] = 1
```
Comment (ritvikvasan) [inline cyto_dl/image/transforms/generate_jepa_masks.py:93]: do you not randomly sample dims in Z here?

File: cyto_dl/models/jepa/iwm.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        source = batch[self.hparams.source_key]
+        source = self.remove_first_dim(source)
+        target = batch[self.hparams.target_key]
+        target = self.remove_first_dim(target)
+
+        # use predictor to predict each patch
+        target_masks = self.get_predict_masks(source.shape[0], device=source.device)
+
+        # mean across patches, no cls token to remove
+        source_embeddings = self.encoder(source)
```
Comment (ritvikvasan) [inline cyto_dl/models/jepa/iwm.py:110]: is the encoder here a vit? Does it handle the patchifying etc? Is the source here the small mask?

File: cyto_dl/models/jepa/jepa_base.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            target_embeddings = self.teacher(x)
+        target_embeddings = rearrange(target_embeddings, "b t c -> t b c")
+        target_embeddings = take_indexes(target_embeddings, mask)
+        target_embeddings = rearrange(target_embeddings, "t b c -> b t c")
+        return target_embeddings
+
+    def get_context_embeddings(self, x, mask):
+        # mask context pre-embedding to prevent leakage of target information
+        context_patches, _, _, _ = self.encoder.patchify(x, 0)
+        context_patches = take_indexes(context_patches, mask)
```
Comment (ritvikvasan) [inline cyto_dl/models/jepa/jepa_base.py:107]: take_indexes seems to be repeating indices along a channel dimension. just want to check that this is expected

File: cyto_dl/nn/encoder_decoder.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        self.encoder = self._init_model(encoder, encoder_ckpt, train_encoder)
+        self.decoder = self._init_model(decoder, decoder_ckpt, train_decoder)
+
+        self.latent_norm = latent_norm if latent_norm is not None else torch.nn.Identity()
+
+    def _init_model(self, model, ckpt, train):
+        if ckpt is not None:
+            state_dict = torch.load(ckpt)["state_dict"]
+            state_dict = {
+                k.replace("backbone.", ""): v
```
Comment (ritvikvasan) [inline cyto_dl/nn/encoder_decoder.py:26]: can you add a comment about why this is necessary

File: cyto_dl/nn/vits/jepa.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+                    encoder_dim=emb_dim,
+                    decoder_dim=emb_dim,
+                    num_heads=num_head,
+                )
+                for _ in range(num_layer)
+            ]
+        )
+
+        self.mask_token = torch.nn.Parameter(torch.zeros(1, 1, emb_dim))
+        self.pos_embedding = torch.nn.Parameter(torch.zeros(np.prod(num_patches), 1, emb_dim))
```
Comment (ritvikvasan) [inline cyto_dl/nn/vits/jepa.py:107]: maybe use the general pos embedding function?

**Review by ritvikvasan — APPROVED** [review]
Looks good to me...Left some minor comments again

**Comment by ritvikvasan** [comment]
general comment: I think having a quick lookup table of the different configs you are adding would be great. Something like this - https://github.com/MMV-Lab/mmv_im2im/blob/main/tutorials/example_by_use_case.md

File: cyto_dl/image/transforms/generate_jepa_masks.py (hunk 6 above)
Comment (benjijamorris) [reply cyto_dl/image/transforms/generate_jepa_masks.py:93]: no - I'm basing this off of v-jepa (which has masks that are the same over time) with the idea that 3d images have similar spatial redundancy to natural videos

File: cyto_dl/models/jepa/iwm.py (hunk 7 above)
Comment (benjijamorris) [reply cyto_dl/models/jepa/iwm.py:110]: yes, the vit encoder does patchifying. source is an image. This is taking all the patches from the source image and predicting the embeddings of all the patches from the target image.

File: cyto_dl/nn/vits/jepa.py (hunk 10 above)
Comment (benjijamorris) [reply cyto_dl/nn/vits/jepa.py:107]: updated!

**Comment by benjijamorris** [comment]
agreed, I'll do a separate pr with that for all the models

File: cyto_dl/image/transforms/generate_jepa_masks.py (hunk 6 above)
Comment (ritvikvasan) [reply cyto_dl/image/transforms/generate_jepa_masks.py:93]: gotcha

File: cyto_dl/models/jepa/jepa_base.py (hunk 8 above)
Comment (benjijamorris) [reply cyto_dl/models/jepa/jepa_base.py:107]: yes, this is just applying the indices across all channels of the tensor (tokens x batch x embedding_dim)

File: cyto_dl/nn/encoder_decoder.py (hunk 9 above)
Comment (benjijamorris) [reply cyto_dl/nn/encoder_decoder.py:26]: I'm thinking this guy isn't ready for prime time yet

**Review by ritvikvasan — APPROVED** [review]
